src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs

```diff
@@ -65,18 +62,52 @@ public override void WaitForCompletion(AwaitAdapter awaiter)
             }
         }
 
-#if NET40 || NET45
         private sealed class WindowsFormsMessagePumpStrategy : MessagePumpStrategy
         {
-            public static readonly WindowsFormsMessagePumpStrategy Instance = new WindowsFormsMessagePumpStrategy();
-            private WindowsFormsMessagePumpStrategy() { }
+            private static WindowsFormsMessagePumpStrategy _instance;
+
+            private readonly Action _applicationRun;
+            private readonly Action _applicationExit;
+
+            private WindowsFormsMessagePumpStrategy(Action applicationRun, Action applicationExit)
+            {
+                _applicationRun = applicationRun;
+                _applicationExit = applicationExit;
+            }
+
+            public static MessagePumpStrategy GetIfApplicable()
+            {
+                if (!IsApplicable(SynchronizationContext.Current)) return null;
+
+                if (_instance is null)
+                {
+                    var applicationType = SynchronizationContext.Current.GetType().Assembly.GetType("System.Windows.Forms.Application", throwOnError: true);
+
+                    var applicationRun = (Action)applicationType
+                        .GetMethod("Run", BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly, null, Type.EmptyTypes, null)
+                        .CreateDelegate(typeof(Action));
+
+                    var applicationExit = (Action)applicationType
+                        .GetMethod("Exit", BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly, null, Type.EmptyTypes, null)
+                        .CreateDelegate(typeof(Action));
+
+                    _instance = new WindowsFormsMessagePumpStrategy(applicationRun, applicationExit);
```

Comment:
There is no need for racing threads to agree upon a single WindowsFormsMessagePumpStrategy instance, so Interlocked.CompareExchange wouldn't add value.